lib/kafka_producer.js

```diff
@@ -91,18 +105,91 @@ KafkaProducer.prototype.connect = function connect(onConnect) {
 
 KafkaProducer.prototype.produce = function produce(topic, message, timeStamp, callback) {
     var self = this;
+
     if (self.restClient) {
-        var produceMessage = self.getProduceMessage(topic, message, timeStamp, 'binary');
-        self.restClient.produce(produceMessage, function handleResponse(err, res) {
-            if (callback) {
-                callback(err, res);
-            }
-        });
+        if (self.batching) {
+            self.batch(topic, message, timeStamp, callback);
+        } else {
+            var produceMessage = self.getProduceMessage(topic, message, timeStamp, 'binary');
+            self.restClient.produce(produceMessage, function handleResponse(err, res) {
+                if (callback) {
+                    callback(err, res);
+                }
+            });
+        }
     } else if (callback) {
         callback(new Error('Kafka Rest Client is not initialized!'));
     }
 };
 
+// Add message to topic's BatchMessage in cache or add new topic to cache
+// Flush topic's BatchMessage on flushCycleSecs interval or if greater than maxBatchSizeBytes
+KafkaProducer.prototype.batch = function batch(topic, message, timeStamp, callback) {
+    var self = this;
+    var currTime = new Date().getTime();
+    if (topic in self.topicToBatchQueue) {
+        var queue = self.topicToBatchQueue[topic];
+        if ((queue.messageBatch.sizeBytes + message.length) > self.maxBatchSizeBytes ||
```

Comment:
nitpick: `message.length` is either number of characters or bytes depending on string vs buffer.

Be wary that `string.length` and `Buffer.byteLength(string)` are different ( multi-byte utf8 characters ). For this particular case we can use `str.length` because its faster then `Buffer.byteLength(x)` and accuracy is not that important.